windows/CodePushPackage.cs

```diff
@@ -0,0 +1,364 @@
+                await downloadFile.DeleteAsync();
+
+                // Merge contents with current update based on the manifest
+                StorageFile diffManifestFile = null;
+                try
+                {
+                    diffManifestFile = await unzippedFolder.GetFileAsync(DIFF_MANIFEST_FILE_NAME);
+                }
+                catch (FileNotFoundException)
+                {
+                    // There is no diff manifest, so this is not a diff update.
+                }
+
+                if (diffManifestFile != null)
+                {
+                    StorageFolder currentPackageFolder = await GetCurrentPackageFolder();
+                    if (currentPackageFolder == null)
+                    {
+                        throw new CodePushInvalidUpdateException("Received a diff update, but there is no current version to diff against.");
+                    }
+
+                    await CodePushUpdateUtils.CopyNecessaryFilesFromCurrentPackage(diffManifestFile, currentPackageFolder, newUpdateFolder);
+                    await diffManifestFile.DeleteAsync();
+                }
+
+                await FileUtils.MergeDirectories(unzippedFolder, newUpdateFolder);
+                await unzippedFolder.DeleteAsync();
+
+                // For zip updates, we need to find the relative path to the jsBundle and save it in the
+                // metadata so that we can find and run it easily the next time.
+                string relativeBundlePath = await CodePushUpdateUtils.FindJSBundleInUpdateContents(newUpdateFolder, expectedBundleFileName);
+                if (relativeBundlePath == null)
+                {
+                    throw new CodePushInvalidUpdateException("Update is invalid - A JS bundle file named \"" + expectedBundleFileName + "\" could not be found within the downloaded contents. Please check that you are releasing your CodePush updates using the exact same JS bundle file name that was shipped with your app's binary.");
+                }
+                else
+                {
+                    if (diffManifestFile != null)
+                    {
+                        // TODO verify hash for diff update
+                        // CodePushUpdateUtils.verifyHashForDiffUpdate(newUpdateFolderPath, newUpdateHash);
+                    }
+
+                    updatePackage[RELATIVE_BUNDLE_PATH_KEY] = relativeBundlePath;
+                }
+            }
+            catch (InvalidDataException)
+            {
+                // Downloaded file is not a zip, assume it is a jsbundle
+                await downloadFile.RenameAsync(expectedBundleFileName);
+                await downloadFile.MoveAsync(newUpdateFolder);
+            }
+
+            // Save metadata to the folder
+            await FileIO.WriteTextAsync(newUpdateMetadataFile, JsonConvert.SerializeObject(updatePackage));
+        }
+
+        public async Task InstallPackage(JObject updatePackage, bool removePendingUpdate)
```

Comment:
I think this method needs some comments to be a little clearer. From now, what I understand is "if removePendingUpdate, then the current package is removed from the current package info and replaced by the update package and the previous package remains--otherwise, the previous package is removed and replaced by the current package and the update package is put as the current package." The phrase "removePendingUpdate" is not descriptive enough for me because what it's really doing is deciding which package remains in the package info, and while the current package is assumed to be a pending update, that is not inforced or explained in this code anywhere.